### PR TITLE
feat(asset): create delete assets API in compass

### PR DESCRIPTION
**Pull request by luthfifahlevi**
**Background**

Currently, there are two services for data discovery: Meteor and Compass. **Meteor** extracts assets from various sources (e.g., BigQuery tables, Optimus, Firehose, Dagger jobs) and syncs the metadata to **Compass** as the catalog service. However, currently, Compass has a lot of assets that no longer present and not synced. **We need a housekeeping mechanism in Compass to automatically remove stale assets from the database.**

**Goals in this Repo:**
Create proto for delete assets API in Compass

**Idea:**
`DeleteAssets` as DELETE `/v1/assets/`: an async API that will return how many rows that will affect based on filter criteria, and delete those rows from PostgreSQL (assets and lineage) and Elasticsearch in the background.

**Request:**
- `query_expr` represent query expr for filter criteria: [https://github.com/expr-lang/expr](https://github.com/expr-lang/expr).
- `dry_run` represent whether it only return how much affected rows (if set to true), or also perform deletion process in the background (default: false).

**Response:**
`affected_rows` represent how many rows that will affect based on filter criteria.

### CODE REVIEW

File: java/build.gradle
Comment (sumslim) [inline java/build.gradle:50]: please remove this before final merging

File: java/build.gradle
Comment (sumslim) [inline java/build.gradle:59]: same for this

File: gotocompany/compass/v1beta1/service.proto
Comment (sumslim) [inline gotocompany/compass/v1beta1/service.proto:1075]: where does the equality condition fit in?

File: gotocompany/compass/v1beta1/service.proto
Comment (sumslim) [inline gotocompany/compass/v1beta1/service.proto:1204]: should be `min_len` or `ignore_empty`?

File: gotocompany/assets/v1beta2/asset.proto
Comment (irainia) [inline gotocompany/assets/v1beta2/asset.proto:65]: change it to `refreshed` instead of `synced`

File: gotocompany/compass/v1beta1/service.proto
Comment (irainia) [inline gotocompany/compass/v1beta1/service.proto:1072]: why do we need the changes in the following line?

File: java/build.gradle
Comment (luthfifahlevi) [reply java/build.gradle:50]: oh, I think that's okay since it beneficial to build on local..? Why need to remove mas?

File: gotocompany/compass/v1beta1/service.proto
Comment (irainia) [inline gotocompany/compass/v1beta1/service.proto:1181]: do we need all the fields here? can we reduce it further if it's not really required at the moment?

**Review by irainia — CHANGES_REQUESTED** [review]

File: gotocompany/compass/v1beta1/service.proto
Comment (luthfifahlevi) [reply gotocompany/compass/v1beta1/service.proto]: Currently, I don't including equality condition. Btw, that current name should be duration_for_upper_threshold.
What do you think about that? Is it better "<=" instead of "<"?

File: gotocompany/compass/v1beta1/service.proto
Comment (luthfifahlevi) [reply gotocompany/compass/v1beta1/service.proto:1210]: Currently, the idea is adding time comparing to Filter type in Compass, and in DeleteAssetsRequest, the Filter will be required, with duration_for_upper_threshold (duration_before_current before) is required (not empty), thats why I set min_len

File: gotocompany/compass/v1beta1/service.proto
Comment (luthfifahlevi) [reply gotocompany/compass/v1beta1/service.proto:1181]: Currently, my idea is Delete Assets will use Filter type in Compass (same as GetAllAssets), with mandatory filled for duration_for_upper_threshold field.

So, I just make it similar to GetAllAssetsRequest, except unnecessary field for DeleteAssets (e.g, size, sort).

File: gotocompany/compass/v1beta1/service.proto
Comment (luthfifahlevi) [reply gotocompany/compass/v1beta1/service.proto:1072]: Currently, my idea is Delete Assets will use **Filter** type in Compass (same as GetAllAssets), with mandatory filled for duration_for_upper_threshold field.

So, I think it is okay to make GetAllAssets also can filter by time (refreshed_at, updated_at, or created_at), since use same Filter type in Compass.

File: gotocompany/compass/v1beta1/service.proto
Comment (luthfifahlevi) [reply gotocompany/compass/v1beta1/service.proto:1181]: Any concern mbak regarding that?

File: gotocompany/compass/v1beta1/service.proto
Comment (haveiss) [inline gotocompany/compass/v1beta1/service.proto:1193]: i don't think we need this for now

Comment (haveiss) [reply gotocompany/compass/v1beta1/service.proto:1193]: instead, we need filter like this
https://github.com/goto/proton/blob/64dd189a923a3474f3ece43542313ee264430a2b/gotocompany/compass/v1beta1/service.proto#L837-L839

File: gotocompany/compass/v1beta1/service.proto
Comment (haveiss) [reply gotocompany/compass/v1beta1/service.proto:1072]: let's skip this changes for now? if needed can create separate PR

File: gotocompany/compass/v1beta1/service.proto
Comment (luthfifahlevi) [reply gotocompany/compass/v1beta1/service.proto:1193]: Okay, I will remove q and q_fields ya, and It is similar to `map<string, string> data = 5`, but I will replace as your mention code 👍

File: gotocompany/compass/v1beta1/service.proto
Comment (haveiss) [reply gotocompany/compass/v1beta1/service.proto:1072]: if we plan to use duration, let's use `google.protobuf.Duration` instead of `string`

File: gotocompany/compass/v1beta1/service.proto
Comment (haveiss) [inline gotocompany/compass/v1beta1/service.proto:1209]: if we plan to use duration, let's use `google.protobuf.Duration` instead of `string`

File: gotocompany/compass/v1beta1/service.proto
Comment (haveiss) [reply gotocompany/compass/v1beta1/service.proto:1210]: use required instead?
https://github.com/bufbuild/protoc-gen-validate?tab=readme-ov-file#durations

File: gotocompany/compass/v1beta1/service.proto
Comment (luthfifahlevi) [inline gotocompany/compass/v1beta1/service.proto:1205]: Are we still need types and services mas @haveiss if this filter exists?
https://github.com/goto/proton/blob/64dd189a923a3474f3ece43542313ee264430a2b/gotocompany/compass/v1beta1/service.proto#L837-L839

What I know that types and service for filtering in multiple items.

File: gotocompany/compass/v1beta1/service.proto
Comment (luthfifahlevi) [reply gotocompany/compass/v1beta1/service.proto:1072]: okayy

File: gotocompany/compass/v1beta1/service.proto
Comment (luthfifahlevi) [reply gotocompany/compass/v1beta1/service.proto]: still needed 👍

File: gotocompany/compass/v1beta1/service.proto
Comment (luthfifahlevi) [reply gotocompany/compass/v1beta1/service.proto]: Okay, updated

File: gotocompany/compass/v1beta1/service.proto
Comment (luthfifahlevi) [reply gotocompany/compass/v1beta1/service.proto]: updated 👍

File: gotocompany/compass/v1beta1/service.proto
Comment (luthfifahlevi) [reply gotocompany/compass/v1beta1/service.proto:1164]: Need your approval since you request changes in here 😅 🙏

Comment (luthfifahlevi) [reply gotocompany/compass/v1beta1/service.proto:1164]: oh sorry mas Anwar, so I thought irainia was another person (woman) 😆

File: gotocompany/compass/v1beta1/service.proto
Comment (batrov) [inline gotocompany/compass/v1beta1/service.proto:1197]: for now can add the filter only for refreshed_at

File: gotocompany/compass/v1beta1/service.proto
Comment (mabdh) [inline gotocompany/compass/v1beta1/service.proto:349]: I am still unsure whether having DELETE with query params is a good idea or not.
Or should we just use POST with req body for this.

Comment (luthfifahlevi) [reply gotocompany/compass/v1beta1/service.proto:349]: After talking again, we decided to send a POST request with a body, following deletion API in elastic search https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-delete-by-query.html

Thank you mas 🙏

File: gotocompany/compass/v1beta1/service.proto
Comment (mabdh) [inline gotocompany/compass/v1beta1/service.proto:1179]: should we add a field if the operation is `dry_run` or not? for the visibility to the user
and we will make the `dry_run` true by default right?

Comment (luthfifahlevi) [reply gotocompany/compass/v1beta1/service.proto:1179]: By discussion with mas Haveiss, `dry_run` is false by default, so it really performs deletion if not specified.
And in response, we don't think it necessary because the client should know about it since he set it as true or false.

**Review by mabdh — APPROVED** [review]

File: gotocompany/assets/v1beta2/asset.proto
Comment (anjali9791) [inline gotocompany/assets/v1beta2/asset.proto:67]: IMO, Currently if nothing is changed in the Asset then we don't send an update call to the compass DB, due to the introduction of this flag, even if nothing else changed we will cause an update to the DB. This will result in much more update at the time of asset update/create, is that okay?

Comment (utsav14nov) [inline gotocompany/assets/v1beta2/asset.proto:67]: Is refreshed_at required? Can't it we handle through `update_time` ?

Comment (luthfifahlevi) [reply gotocompany/assets/v1beta2/asset.proto:67]: Of course there are some advantages and drawbacks, but I think that's still okay, since we also reduce the cost of saving the stale assets 🙏  CMIIW mas @haveiss

Comment (luthfifahlevi) [reply gotocompany/assets/v1beta2/asset.proto:67]: Yap, It's capable of handling situations where an asset doesn't receive any single updates, although it already synced/refreshed (should **not** categorize as a stale asset), which is something we cannot detect with `updated_at` field. That's why we introduce that flag

**Pull request merged**
